### ISSUE TITLE
Too many include directories in compile command

### ISSUE THREAD

**Issue by kjellahl** [mesonbuild/meson#8562]
The `implicit_include_directories: false` kwarg in library() sometimes, but not
always, correctly excludes the current source and build directories from the
include path. An exception is that the current build directory is not excluded
if there are generated files (custom_target()) in that directory, and those generated
files are sources of library(). I suspect that these lines in
mesonbuild/backend/ninjabackend.py are relevant.
```
        # Add custom target dirs as includes automatically, but before
        # target-specific include directories.
        # XXX: Not sure if anyone actually uses this? It can cause problems in
        # situations which increase the likelihood for a header name collision,
        # such as in subprojects.
        commands += self.get_custom_target_dir_include_args(target, compiler)
```
The problem that is foreseen in the comment has come true.

The libsigc++2 library generates a file called `signal.h` in the `sigc++` directory.
A source code file that uses sigc++/signal.h contains
```
#include <sigc++/signal.h>
```
That's no problem when other projects use an installed version of libsigc++2.
When libsigc++ is a subproject, the unwanted include directory is a problem.
A source code file in a superproject that includes the system file `signal.h` with
```
#include <signal.h>
```
finds `sigc++/signal.h`.

How can I avoid the unwanted `-Isigc++` argument in compile commands?

meson version: 0.57.1

**Comment by jpakkane** [comment]
At the moment that is not possible, sadly. We'd probably need to change it so that `implicit_include_directories` also covers the one coming from custom_target.

**Comment by jpakkane** [comment]
Can you try if the linked MR fixes the issue for you?

**Comment by kjellahl** [comment]
The patch in the MR fixes only a small part of the issue.

1. Outputs from custom_target() in a subproject are inputs to library().
The unwanted include directory *is* removed there.

2. The output from library() is input to `link_with:` in declare_dependency().
The output from declare_dependency() is a dependency of executable() in
the subproject. The unwanted include directory *is not* removed there.

3. The output from declare_dependency() is a dependency of library() and executable()
in the superproject. The unwanted include directory *is not* removed there.

Step 3 is the important step for me.

**Comment by jpakkane** [comment]
[](I tried to recreate this but could not do it. A sample project is attached.

[deptest.zip](https://github.com/mesonbuild/meson/files/6213345/deptest.zip)

Opening the generated `build.ninja` it is easy to see that the subdirectory path is not added to the executable's include paths. It does get added if you remove the `implicit_include_directories` kwarg from it.

**Comment by kjellahl** [comment]
So I put `implicit_include_directories: false` in the wrong meson.build file?
Or did you? deptest2.zip shows how I thought it should be done:
`implicit_include_directories: false` in static_library() in
deptest2/subprojects/sp/meson.build, and only there.

A subproject can be an alternative to one of pkg-config's .pc files.
`dep1 = dependency()` can get information either from a .pc file or from
declare_dependency() (+ meson.override_dependency()) in a subproject. A .pc file
shows which directories shall be searched for include files, belonging to the
dependency. No extra kwarg is needed in an executable() or a library() that
depends on dep1 in order to exclude unwanted search paths. I think it's reasonable
to expect that it should work accordingly when the information comes from a
subproject.

I noticed that implicit_include_directories is unnecessarey in deptest2/meson.build
if I remove `sources: genh` from `d = declare_dependency(sources: genh, link_with: l)`
in deptest2/subprojects/sp/meson.build. But, unless I have misunderstood, that would
mean that .c files that include the generated header file can be compiled before
the header file has been generated.

It looks like the `sources` kwarg in declare_dependency() has two meanings:
Don't compile before these files have been generated, and search for include
files in the directories where these files are stored. There ought to be a way
of specifying in declare_dependency() that only the first meaning is wanted.

[deptest2.zip](https://github.com/mesonbuild/meson/files/6215604/deptest2.zip)

**Comment by kjellahl** [comment]
The fix in commit https://github.com/mesonbuild/meson/commit/e80ff985fb1d4a0b840e5bf67a7e5dff08a21cdd
solves my immediate problem. Problems with finding the wrong header file can be avoided
by adding `implicit_include_directories: false` in some calls to library() and executable().

The docs/markdown/snippets/customimplicit.md file that was added by the commit
hints at a better solution. It says: *the directory will no longer be put in the
search path if the target has `implicit_include_directories: false`*. Which target?
The custom_target()? No, `implicit_include_directories: false` goes into the executable().
It would be better if it was possible to specify in custom_target() that the output
directory shall not be automatically added to any header search path.
It would be especially advantageous when the custom_target() is in a subproject,
and the generated header file is used in a superproject. The developers of the
superproject would not have to know that they must add `implicit_include_directories: false`
in library() and executable() in the superproject just to avoid adding too many
search paths from the subproject.

**Comment by kjellahl** [comment]
Fixed in Meson 0.58.0. The fix in commit e80ff985fb1d4a0b840e5bf67a7e5dff08a21cdd is good enough.
